dll/win32/shell32/propsheet.cpp

```diff
@@ -0,0 +1,178 @@
+}
+
+static HRESULT
+SHELL_InitializeExtension(REFCLSID clsid, PCIDLIST_ABSOLUTE pidlFolder, IDataObject *pDO, HKEY hkeyProgID, REFIID riid, void **ppv)
+{
+    *ppv = NULL;
+    IUnknown *pUnk;
+    HRESULT hr = SHCoCreateInstance(NULL, &clsid, NULL, riid, (void**)&pUnk);
+    if (SUCCEEDED(hr))
+    {
+        CComPtr<IShellExtInit> Init;
+        hr = pUnk->QueryInterface(IID_PPV_ARG(IShellExtInit, &Init));
+        if (SUCCEEDED(hr))
+            hr = Init->Initialize(pidlFolder, pDO, hkeyProgID);
+
+        if (SUCCEEDED(hr))
+            *ppv = (void*)pUnk;
+        else
+            pUnk->Release();
+    }
+    return hr;
+}
+
+static HRESULT
+AddPropSheetHandlerPages(REFCLSID clsid, IDataObject *pDO, HKEY hkeyProgID, PROPSHEETHEADERW &psh)
+{
+    CComPtr<IShellPropSheetExt> SheetExt;
+    HRESULT hr = SHELL_InitializeExtension(clsid, NULL, pDO, hkeyProgID, IID_PPV_ARG(IShellPropSheetExt, &SheetExt));
+    if (SUCCEEDED(hr))
+    {
+        UINT OldCount = psh.nPages;
+        hr = SheetExt->AddPages(AddPropSheetPageCallback, (LPARAM)&psh);
+        if (hr > 0) // The returned index is one-based (relative to this extension).
+        {
+            hr += OldCount;
+            psh.nStartPage = hr - 1;
+        }
+    }
+    return hr;
+}
+
+static HRESULT
+SHELL_CreatePropSheetStubWindow(CStubWindow32 &stub, PCIDLIST_ABSOLUTE pidl, const POINT *pPt)
+{
+    PWSTR Path;
+    if (!pidl || FAILED(SHGetNameFromIDList(pidl, SIGDN_DESKTOPABSOLUTEPARSING, &Path)))
+        Path = NULL; // If we can't get a path, we simply will not be able to reuse this window
+
+    HRESULT hr = CStubWindow32::CreateStub(stub, CStubWindow32::TYPE_PROPERTYSHEET, Path, pPt);
+    SHFree(Path);
+    UINT flags = SHGFI_ICON | SHGFI_ADDOVERLAYS;
+    SHFILEINFO sfi;
+    if (SUCCEEDED(hr) && SHGetFileInfoW((LPWSTR)pidl, 0, &sfi, sizeof(sfi), SHGFI_PIDL | flags))
+        stub.SendMessage(WM_SETICON, ICON_BIG, (LPARAM)sfi.hIcon);
+    return hr;
+}
+
+/*************************************************************************
+ *  SHELL32_PropertySheet [INTERNAL]
+ *  PropertySheetW with stub window.
+ */
+static INT_PTR
+SHELL32_PropertySheet(LPPROPSHEETHEADERW ppsh, IDataObject *pDO)
+{
+    CStubWindow32 stub;
+    POINT pt, *ppt = NULL;
+    if (pDO && SUCCEEDED(DataObject_GetOffset(pDO, &pt)))
+        ppt = &pt;
+    PIDLIST_ABSOLUTE pidl = SHELL_DataObject_ILCloneFullItem(pDO, 0);
+    HRESULT hr = SHELL_CreatePropSheetStubWindow(stub, pidl, ppt);
+    ILFree(pidl);
+    if (FAILED(hr))
+        return hr == HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS) ? 0 : -1;
+
+    INT_PTR Result = -1;
+    ppsh->hwndParent = stub;
+    if (ppsh->nPages)
+    {
+        Result = PropertySheetW(ppsh);
+    }
+    else
+    {
+        WCHAR szFormat[128], szMessage[_countof(szFormat) + 42];
+        LoadStringW(shell32_hInstance, IDS_CANTSHOWPROPERTIES, szFormat, _countof(szFormat));
+        wsprintfW(szMessage, szFormat, ERROR_CAN_NOT_COMPLETE);
```

Comment:
StringCchPrintfW

Reply:
We don't have helpers for showing such errors (with printf-like support) already in shell32?

Reply:
There is `ShellMessageBoxW` but that uses `FormatMessageW` and is not compatible with `IDS_CANTSHOWPROPERTIES`. All these `IDS_CANTxyz` strings use printf style strings.

Feel free to fix this and all other strings like it in another PR.